### DISCUSSION TITLE
IKEv2 tunnel fail to rekey on IPv6, but never for IPv4

### DISCUSSION THREAD

**Discussion by SvenRoederer** [strongswan/strongswan#2399] [Q&A]
I use a roadwarrior configuration with split-tunnel and MOBIKE based on IKEv2.  Gateway runs OpenWrt (StrongSwan 5.9.10, also tested with 5.9.14) and is connected to our ISP-line with a public IPv4 address. Clients are running NCP EntryClient v13.14 and Windows10 IpSec client and are usually behind some NAT of employees router at home.
I have a really stable setup where employees can connect to company-LAN without issue. The tunnel will stay up for more tan the usual work-hours. If I remember correctly, tunnel and IKE-session will break when ISP forces reconnect and assigns new addresses to the client. 

Some time ago I got public IPv6 and added this in Dualstack setup. So the gateway has public IPv4 and IPv6 address.  After this change I get user reports, that the connection fails. it'S not the IKE-session, but the ESP-tunnel. This happens usually after 3,5h to 4,5h.
It turns out, that only Clients connecting via IPv6 see the behavior, IPv4-clients are still fine. Esp. in NCP EntryClient I can see in the logs, that keepalive is successfully exchanged between gateway and client all the time, also a valid SA seems to be present, but there is not kind of traffic possible inside the tunnel.
As I have enabled MOBIKE the tunnel is always ESP-over-UDP, as outlined in https://docs.strongswan.org/docs/5.9/features/mobike.html.

My config for swancrt:

~~~
# taken from https://www.strongswan.org/testing/testresults/swanctl/ip-pool/

connections {

   CVT-vpn-XFRM {
      local_addrs  = 77.245.36.50 , 2a03:5c40:8::d1a7:0:0
      pools = rw_pool
      if_id_in = 42
      if_id_out = 42

      local {
         auth = pubkey
         certs = /etc/swanctl/x509/vpn-gateway.crt
         id = "xxxx"
      }
      remote {
         auth = eap-tls
         eap_id = %any
         cacerts = /etc/swanctl/x509ca/ColVisTec_VPN-CAr2.crt
      }
      children {
         net {
            local_ts = 172.16.a.b/24
            esp_proposals = aes128-sha2_256-modp2048
            updown = /etc/colvistec-vpn/xfrm.sh
         }
      }
      version = 2
   }
}

pools {
   rw_pool {
      addrs = 172.16.x.y-172.16.x.z
   }
}

authorities {
   ColVisTec {
      cacert = /etc/swanctl/x509ca/ColVisTec_Root-CA.crt
      crl_uris = http://****.intern.colvistec.de/*****
   }
}
~~~

The updown-script is just build around:

~~~
        [ -z "`ip rule show table ${TABLE}`" ] && ip rule add table ${TABLE} prio 220
        ip route add ${PLUTO_PEER_CLIENT} dev "${VTI_IF}" table ${TABLE}
~~~

Recently I found some time to debugged a bit on this issue. 
It looks to me, that the ESP-tunnel stalls when the IKE-session gets rekeyed. Rekeying of SA is working for both kinds of IP-transport. 

What I found: I disabled MOBIKE-suppport on the gateway and expected to get native ESP-traffic, but still got ESP-over-UDP. This resulted from NAT-D situation detected on IPv6. When the connection is negotiated the logs of strongswan show that "NAT-D" is found. I wonder how, as mention the gateway is directly attached to public-IPaddress and my client at home is also not treated with NAT for IPv6. Late the log shows, when rekeying is done, that NAT-D suddenly disappeared. There was no active change at any side of the network-equipment/hardware/configuration.

~~~
08.08.2024 23:11:49 - Ikev2Info: ConRef=28, SendNotifyMsgUpdateSaAddress
08.08.2024 23:11:49 - Ikev2: Turning on NATD mode - CVT-ExtDev - 1                                      <<-----that's missing when rekeying happens
08.08.2024 23:11:49 - IPSec: Quick Mode is Ready: IkeIndex=28,VpnSrcPort=10954
...
09.08.2024 02:13:39 - Ikev2Info: ConRef=28, SendNotifyMsgUpdateSaAddress
09.08.2024 02:13:39 - IPSec: Quick Mode is Ready: IkeIndex=28,VpnSrcPort=10954
~~~

For the initial key-setup client and gateway agree on NAT-D (Gateway: `Turning on NATD mode` / client:`Gateway=2a03:5c40:0008:0000:0000:defe:0109:0000,NatdMode=1,Roamingcon=1`) , but for Rekey they disagree (Gateway: "No NAT-D" / client: `Gateway=2a03:5c40:0008:0000:0000:defe:0109:0000,NatdMode=1,Roamingcon=1`).

See also the full logs: [2024-08-08_22-52_NCPclient-log.txt](https://github.com/user-attachments/files/16599952/2024-08-08_22-52_NCPclient-log.txt), [2024-08-08_21-16_strongswan.txt](https://github.com/user-attachments/files/16599957/2024-08-08_21-16_strongswan.txt)

I found that settnig 'encap=yes" on the connection seems to fix the problem. That's not really unexpected, as NAT-D will be used always. For IPv6 there should actually be no need and for IPv4 it's used and was used always, as NAT is done for real.

Following questions:
* Why both parties initially detect NAT on the link?
* Why during Rekey they don't agree in NAT?
* If IKE NAT-D mismatch is the cause, I expect strongSwan as the potential cause, as I can see this for NCP client IpSec-implementation and also on Windows10 native IpSec.

Anyone has seen something similar?

**Comment by tobiasbrunner** [comment]
> Why both parties initially detect NAT on the link?

If there is no NAT, it might be forced by one of the implementations (sending random NAT detection notifies, which is what `encap=yes` does, it does nothing else). What exactly is logged?

> Why during Rekey they don't agree in NAT?

It's not (re-)negotiated during rekeying. It could be during MOBIKE exchanges (i.e. if NAT-D payloads are added to DPDs), but otherwise it's a property of the IKE_SA that does not suddenly change. So this might be a bug in the other implementation. However, note that the Linux kernel does not support processing non-encapsulated ESP packets on SAs that have UDP-encapsulation enabled and vice-versa, which is technically a violation of the RFCs. Forcing UDP encap (i.e. `encap=yes`) could help if a peer uses UDP-encap even if there is no NAT.

**Reply by SvenRoederer** [reply]
> > Why both parties initially detect NAT on the link?
> 
> If there is no NAT, it might be forced by one of the implementations (sending random NAT detection notifies, which is what `encap=yes` does, it does nothing else). What exactly is logged?

As mentioned: initially both parties agree on NAT, but when recreating CHIL_SA around "Fri Aug  9 02:13:39 2024" they disagee. StrongSwan `daemon.info ipsec: 11[IKE] remote host is not behind NAT anymore`; NCP-client: `Gateway=xxx,NatdMode=1,Roamingcon=1`
Do you need some other log verbosity?

**Reply by tobiasbrunner** [reply]
Yes, please post the **complete** log.

**Reply by SvenRoederer** [reply]
Current logs include the logging from initial "connection setup (IKE_SA_INIT)" `Aug  8 21:16:41 2024` till some minutes after connection has died `Aug  9 02:13:39 2024` and connection was taken down `Aug  9 02:16:43 2024`. Above logs are for both peers StongSwan and NCPClient.
In addition I can activate "Verbos logging on NCPclient". Shall I include loglevel=4 for all subsystems on StrongSwan?

**Reply by tobiasbrunner** [reply]
Ah, OK. As far as I can see, the NAT detection change coincides with a MOBIKE update from the client, even though the IP addresses and ports have not actually changed. In the client log we see this:

```
09.08.2024 02:13:35 - IPSec: Link Status Address Change Received - IpAdr=192.168.9.173,AdapterIndex=201
09.08.2024 02:13:35 - IPSec: Link Status Address Change Received on Local Tunnel Endpoint - NewIpAdr=192.168.9.173,OldIpadr=2001:09e8:0471:8c81:f907:6f4e:95d0:2993,AdapterIndex=201
09.08.2024 02:13:35 - IPSec: Link Status Address Change Received on Local Tunnel Endpoint with internet connectivity - NewIpAdr=192.168.9.173,OldIpadr=2001:09e8:0471:8c81:f907:6f4e:95d0:2993,AdapterIndex=201
09.08.2024 02:13:38 - osspecific_add_dns: cmdline=netsh interface ipv4 add dnsservers 7 172.16.a.d validate=no
09.08.2024 02:13:39 - Ikev2Info: ConRef=28, SendNotifyMsgUpdateSaAddress
```

So the client seems to think that there is an IP address change on its interface from IPv6 to IPv4. However, the MOBIKE update is still sent from the IPv6 address:

```
Fri Aug  9 02:13:39 2024 daemon.info ipsec: 11[NET] received packet: from 2001:9e8:471:8c81:f907:6f4e:95d0:2993[10954] to 2a03:5c40:8::defe:109:0[4500] (128 bytes)
Fri Aug  9 02:13:39 2024 daemon.info ipsec: 11[ENC] parsed INFORMATIONAL request 6 [ N(UPD_SA_ADDR) N(NATD_D_IP) N(NATD_S_IP) ]
Fri Aug  9 02:13:39 2024 daemon.info ipsec: 11[IKE] remote host is not behind NAT anymore
```

Funnily enough, this message then also seems to have the correct `NATD_S_IP` value that matches the source address and port of the packet (i.e. strongSwan now detects that there is no NAT "anymore"). If you look at the client log, there were several such "address changes" before, as well as corresponding MOBIKE updates. However, the NAT-D value was apparently indicating a NAT all that time.

Anyway, all this seems to indicate a bug/issue on the client.

> Shall I include loglevel=4 for all subsystems on StrongSwan?

No, please always use the log levels given in #196 when debugging. Although for this particular case you should increase the log level for _ike_ to 3 to see details regarding the NAT-D notifies.
